rig/routing_table/ordered_covering.py

```diff
@@ -0,0 +1,752 @@
+    #
+    # In this case there are 2 ``settable`` bits of interest highlighted.
+    # Whenever we have multiple covered entries we care only about the entries
+    # with the fewest number of ``settable`` bits because these most constrain
+    # which entries we may remove from the merge to avoid covering up the lower
+    # entry.
+    #
+    # NOTE:
+    #   * If there is only 1 ``settable`` bit then we are very constrained in
+    #     terms of which entries must be removed from the merge to avoid
+    #     covering a lower entry.
+    #   * If there are no ``settable`` bits then we cannot possibly avoid
+    #     covering the lower entry - the only correct action is to return an
+    #     empty merge.
+    #
+    # Assuming that there were no covered entries with no ``settable`` bits
+    # (that is ``stringency > 0``) then ``bits_and_vals`` contains pairs of
+    # bits and boolean values which indicate which values need to be removed
+    # from which bit positions to avoid covering up lower entries. If the
+    # example above were the only covered entry then ``bits_and_vals`` would
+    # contain ``(m, True)`` to indicate that all entries containing Xs or 1s in
+    # the left-most bit of interest could be removed to avoid the covered entry
+    # and ``(n, False)`` to indicate that all entries containing Xs or 0s in
+    # the right-most bit of interest could be removed to avoid covering the
+    # entry.
+    #
+    # NOTE:
+    #   ``bits_and_vals`` consists of a set of options (e.g., we *could* remove
+    #   all entries with Xs or 1s in bit ``m`` *or* we could remove all entries
+    #   with Xs or 0s in bit ``n``, either would resolve the above covering).
+    #
+    # To determine which course of action to take we build a dictionary mapping
+    # each of the pairs in ``bits_and_vals`` to the entries that would need to
+    # be removed to "set" that bit in the merged entry. For example, we might
+    # end up with:
+    #
+    #     options = {(m, True): {1, 4, 5},
+    #                (n, False): {3, 7}}
+    #
+    # Indicating that we'd need to remove entries 1, 4 and 5 from the merge to
+    # "set" the mth bit of the merged to 0 or that we'd need to remove entries
+    # 3 and 7 to set the nth bit of the merged entry to set the nth bit to 1.
+    #
+    # NOTE:
+    #   The boolean part of the pair indicates which value needs to be removed
+    #   (True -> remove all 1s; False -> remove all 0s). If all Xs and 1s in a
+    #   given bit position are removed from a merge then the merged entry is
+    #   guaranteed to have a 0 in the bit position. Vice-versa removing all Xs
+    #   and 0s in a given bit position from a merge will result in a merged
+    #   entry with a 1 in that position.
+    #
+    # As we want to make our merges as large as possible we select the smallest
+    # set of entries to remove from the merge from ``options``. After removing
+    # these entries we have a new merge with lower generality.
```

Comment:
You can remove this last sentence, you re-state this fact at a more pertinent time in the next paragraph.